File: src/workflow/adapters/code_adapter.py
"""
Code Domain Adapter

Implements a practical code workflow:
planner -> coder -> evaluator -> finalize.
"""

from typing import Type, Any, Dict, Optional
from langgraph.graph import StateGraph, END

from .base_adapter import (
    BaseDomainAdapter,
    AdapterRegistry,
    DomainType,
    BaseEvaluationResult,
)
from src.workflow.base_state import BaseState, create_base_state


@AdapterRegistry.register(
    DomainType.CODE.value,
    capabilities=("strict-governance", "cli-exposed"),
)
class CodeAdapter(BaseDomainAdapter):

    def get_domain_type(self) -> str:
        return DomainType.CODE.value

    def get_state_class(self) -> Type[BaseState]:
        return BaseState

    def create_initial_state(self, user_request: str, **kwargs) -> BaseState:
        metadata = kwargs.get("metadata") or {}
        resume_decision = kwargs.get("resume_decision")
        if resume_decision:
            metadata = {**metadata, "resume_decision": resume_decision}

        return create_base_state(
            user_request=user_request,
            domain=DomainType.CODE.value,
            metadata=metadata,
            **{key: value for key, value in kwargs.items() if key not in ("metadata", "resume_decision")},
        )

    def evaluate(self, state: BaseState) -> BaseEvaluationResult:
        pass_score = float(self.config.get("pass_score", 80))
        coverage_threshold = float(self.config.get("code_coverage_threshold", 80))
        max_revisions = int(self.config.get("max_revisions", 3))

        tests_passed = self._resolve_bool_signal(state, "tests_passed")
        lint_passed = self._resolve_bool_signal(state, "lint_passed")
        build_passed = self._resolve_bool_signal(state, "build_passed")
        coverage = self._resolve_coverage(state)

        dimension_scores = {
            "tests": self._score_binary_signal(tests_passed, 35.0, 18.0),
            "lint": self._score_binary_signal(lint_passed, 20.0, 10.0),
            "build": self._score_binary_signal(build_passed, 25.0, 12.0),
            "coverage": self._score_coverage(coverage),
        }
        total_score = round(sum(dimension_scores.values()), 2)

        errors = state.get("errors") or []
        revision_count = int(state.get("revision_count", 0))
        coverage_below_threshold = coverage is not None and coverage < coverage_threshold

        blocking_reasons = []
        revision_instructions = []

        if errors:
            blocking_reasons.append("runtime_or_execution_errors")
            revision_instructions.append({
                "target": "error",
                "issue": "Resolve execution/runtime errors first",
                "action": "Inspect errors and fix root causes before next evaluation",
            })

        if tests_passed is False:
            blocking_reasons.append("tests_failed")
            revision_instructions.append({
                "target": "tests",
                "issue": "Tests are failing",
                "action": "Fix failing tests and keep regression suite green",
            })

        if lint_passed is False:
            blocking_reasons.append("lint_failed")
            revision_instructions.append({
                "target": "lint",
                "issue": "Lint checks are failing",
                "action": "Apply lint fixes and formatting before merge",
            })

        if build_passed is False:
            blocking_reasons.append("build_failed")
            revision_instructions.append({
                "target": "build",
                "issue": "Build checks are failing",
                "action": "Fix build breakages and verify clean build",
            })

        if coverage_below_threshold:
            blocking_reasons.append("coverage_below_threshold")
            revision_instructions.append({
                "target": "coverage",
                "issue": f"Coverage {coverage:.2f}% below {coverage_threshold:.2f}%",
                "action": "Add/adjust tests to satisfy coverage threshold",
            })

        if blocking_reasons:
            if revision_count >= max_revisions:
                decision = "HUMAN_REVIEW"
                decision_reason = (
                    f"Blocking checks still failing after {revision_count} revisions: "
                    f"{', '.join(blocking_reasons)}"
                )
            else:
                decision = "REVISE"
                decision_reason = f"Blocking checks failed: {', '.join(blocking_reasons)}"
        elif total_score >= pass_score:
            decision = "APPROVED"
            decision_reason = f"Code quality gates passed (score={total_score:.2f})"
        else:
            decision = "REVISE"
            decision_reason = f"Quality score {total_score:.2f} below pass score {pass_score:.2f}"

        feedback = (
            "Code evaluation completed: "
            f"tests={tests_passed}, lint={lint_passed}, build={build_passed}, coverage={coverage}"
        )

        return BaseEvaluationResult(
            decision=decision,
            decision_reason=decision_reason,
            total_score=total_score,
            dimension_scores=dimension_scores,
            feedback=feedback,
            revision_instructions=revision_instructions,
        )

    def create_graph(self):
        workflow = StateGraph(BaseState)
        workflow.add_node("planner", self.planner_node)
        workflow.add_node("coder", self.coder_node)
        workflow.add_node("evaluator", self.evaluator_node)
        workflow.add_node("finalize", self.finalize_node)

        workflow.set_entry_point("planner")
        workflow.add_edge("planner", "coder")
        workflow.add_edge("coder", "evaluator")
        workflow.add_edge("evaluator", "finalize")
        workflow.add_edge("finalize", END)
        return workflow

    async def planner_node(self, state: BaseState) -> Dict[str, Any]:
        user_request = state.get("user_request", "")
        plan = [
            "Analyze requirements and constraints",
            "Implement code changes",
            "Run tests/lint/build/coverage checks",
            "Prepare final output and decision",
        ]
        return {
            "current_step": "planner",
            "context": f"Code task: {user_request}",
            "implementation_plan": plan,
        }

    async def coder_node(self, state: BaseState) -> Dict[str, Any]:
        metadata = state.get("metadata") or {}
        change_summary = metadata.get("change_summary")
        if not change_summary:
            change_summary = f"Implement request: {state.get('user_request', '')}"

        return {
            "current_step": "coder",
            "draft_content": change_summary,
            "iteration_count": int(state.get("iteration_count", 0)) + 1,
        }

    async def evaluator_node(self, state: BaseState) -> Dict[str, Any]:
        evaluation = self.evaluate(state)
        metadata = dict(state.get("metadata") or {})
        metadata["quality_assessment"] = {
            "total_score": evaluation.total_score,
            "dimension_scores": evaluation.dimension_scores,
        }

        return {
            "current_step": "evaluator",
            "decision": evaluation.decision,
            "decision_reason": evaluation.decision_reason,
            "score": evaluation.total_score,
            "feedback_context": evaluation.feedback,
            "revision_instructions": evaluation.revision_instructions,
            "metadata": metadata,
        }

    async def finalize_node(self, state: BaseState) -> Dict[str, Any]:
        decision = state.get("decision", "REVISE")
        approved = decision == "APPROVED"
        return {
            "current_step": "finalize",
            "requires_human_intervention": decision == "HUMAN_REVIEW",
            "final_output": state.get("draft_content", "") if approved else state.get("final_output", ""),
        }

    def _resolve_bool_signal(self, state: BaseState, key: str) -> Optional[bool]:
        value = state.get(key)
        if value is None:
            metadata = state.get("metadata") or {}
            quality_signals = metadata.get("quality_signals") if isinstance(metadata, dict) else {}
            if isinstance(quality_signals, dict) and key in quality_signals:
                value = quality_signals.get(key)
            else:
                value = metadata.get(key) if isinstance(metadata, dict) else None

        return self._to_optional_bool(value)

    def _resolve_coverage(self, state: BaseState) -> Optional[float]:
        coverage_keys = ("coverage", "coverage_pct", "coverage_percent")
        metadata = state.get("metadata") or {}
        quality_signals = metadata.get("quality_signals") if isinstance(metadata, dict) else {}

        value = None
        for key in coverage_keys:
            if key in state:
                value = state.get(key)
                break
            if isinstance(quality_signals, dict) and key in quality_signals:
                value = quality_signals.get(key)
                break
            if isinstance(metadata, dict) and key in metadata:
                value = metadata.get(key)
                break

        if value is None:
            return None

        if isinstance(value, str):
            value = value.strip().rstrip("%")

        try:
            coverage = float(value)
        except (TypeError, ValueError):
            return None

        return max(0.0, min(coverage, 100.0))

    @staticmethod
    def _to_optional_bool(value: Any) -> Optional[bool]:
        if value is None:
            return None
        if isinstance(value, bool):
            return value
        if isinstance(value, (int, float)):
            return bool(value)
        if isinstance(value, str):
            normalized = value.strip().lower()
            if normalized in {"true", "pass", "passed", "yes", "y", "1"}:
                return True
            if normalized in {"false", "fail", "failed", "no", "n", "0"}:
                return False
        return None

    @staticmethod
    def _score_binary_signal(signal: Optional[bool], pass_score: float, unknown_score: float) -> float:
        if signal is True:
            return pass_score
        if signal is False:
            return 0.0
        return unknown_score

    @staticmethod
    def _score_coverage(coverage: Optional[float]) -> float:
        if coverage is None:
            return 8.0
        return round((coverage / 100.0) * 20.0, 2)

File: src-ts/workflow/adapters/code-adapter.ts
/**
 * Code Domain Adapter
 *
 * Implements a practical code workflow:
 * planner -> coder -> evaluator -> finalize.
 *
 * Migrated from src/workflow/adapters/code_adapter.py
 */

import {
  BaseDomainAdapter,
  AdapterRegistry,
  type BaseEvaluationResult,
} from './base-adapter';
import type { BaseState } from '../state';
import { createBaseState } from '../state';

export class CodeAdapter extends BaseDomainAdapter {

  getDomainType(): string {
    return 'code';
  }

  getStateClass(): new (...args: unknown[]) => BaseState {
    return class {} as new (...args: unknown[]) => BaseState;
  }

  createInitialState(userRequest: string, extra: Record<string, unknown> = {}): BaseState {
    let metadata = (extra['metadata'] as Record<string, unknown>) ?? {};
    const resumeDecision = extra['resume_decision'];
    if (resumeDecision) {
      metadata = { ...metadata, resume_decision: resumeDecision };
    }

    const filtered: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(extra)) {
      if (key !== 'metadata' && key !== 'resume_decision') {
        filtered[key] = value;
      }
    }

    return createBaseState(userRequest, {
      domain: 'code',
      metadata,
      ...filtered,
    });
  }

  evaluate(state: BaseState): BaseEvaluationResult {
    const configAny = this.config as Record<string, unknown>;
    const passScore = Number(configAny['pass_score'] ?? 80);
    const coverageThreshold = Number(configAny['code_coverage_threshold'] ?? 80);
    const maxRevisions = Number(configAny['max_revisions'] ?? 3);

    const testsPassed = this._resolveBoolSignal(state, 'tests_passed');
    const lintPassed = this._resolveBoolSignal(state, 'lint_passed');
    const buildPassed = this._resolveBoolSignal(state, 'build_passed');
    const coverage = this._resolveCoverage(state);

    const dimensionScores: Record<string, number> = {
      tests: CodeAdapter._scoreBinarySignal(testsPassed, 35.0, 18.0),
      lint: CodeAdapter._scoreBinarySignal(lintPassed, 20.0, 10.0),
      build: CodeAdapter._scoreBinarySignal(buildPassed, 25.0, 12.0),
      coverage: CodeAdapter._scoreCoverage(coverage),
    };
    const totalScore = Math.round(
      Object.values(dimensionScores).reduce((a, b) => a + b, 0) * 100,
    ) / 100;

    const stateAny = state as Record<string, unknown>;
    const errors = ((stateAny['errors'] as string[]) ?? []);
    const revisionCount = Number(stateAny['revision_count'] ?? 0);
    const coverageBelowThreshold = coverage !== null && coverage < coverageThreshold;

    const blockingReasons: string[] = [];
    const revisionInstructions: Array<Record<string, string>> = [];

    if (errors.length > 0) {
      blockingReasons.push('runtime_or_execution_errors');
      revisionInstructions.push({
        target: 'error',
        issue: 'Resolve execution/runtime errors first',
        action: 'Inspect errors and fix root causes before next evaluation',
      });
    }

    if (testsPassed === false) {
      blockingReasons.push('tests_failed');
      revisionInstructions.push({
        target: 'tests',
        issue: 'Tests are failing',
        action: 'Fix failing tests and keep regression suite green',
      });
    }

    if (lintPassed === false) {
      blockingReasons.push('lint_failed');
      revisionInstructions.push({
        target: 'lint',
        issue: 'Lint checks are failing',
        action: 'Apply lint fixes and formatting before merge',
      });
    }

    if (buildPassed === false) {
      blockingReasons.push('build_failed');
      revisionInstructions.push({
        target: 'build',
        issue: 'Build checks are failing',
        action: 'Fix build breakages and verify clean build',
      });
    }

    if (coverageBelowThreshold) {
      blockingReasons.push('coverage_below_threshold');
      revisionInstructions.push({
        target: 'coverage',
        issue: `Coverage ${coverage!.toFixed(2)}% below ${coverageThreshold.toFixed(2)}%`,
        action: 'Add/adjust tests to satisfy coverage threshold',
      });
    }

    let decision: string;
    let decisionReason: string;

    if (blockingReasons.length > 0) {
      if (revisionCount >= maxRevisions) {
        decision = 'HUMAN_REVIEW';
        decisionReason =
          `Blocking checks still failing after ${revisionCount} revisions: ` +
          `${blockingReasons.join(', ')}`;
      } else {
        decision = 'REVISE';
        decisionReason = `Blocking checks failed: ${blockingReasons.join(', ')}`;
      }
    } else if (totalScore >= passScore) {
      decision = 'APPROVED';
      decisionReason = `Code quality gates passed (score=${totalScore.toFixed(2)})`;
    } else {
      decision = 'REVISE';
      decisionReason = `Quality score ${totalScore.toFixed(2)} below pass score ${passScore.toFixed(2)}`;
    }

    const feedback =
      `Code evaluation completed: ` +
      `tests=${testsPassed}, lint=${lintPassed}, build=${buildPassed}, coverage=${coverage}`;

    return {
      decision,
      decision_reason: decisionReason,
      total_score: totalScore,
      dimension_scores: dimensionScores,
      feedback,
      revision_instructions: revisionInstructions,
    };
  }

  createGraph(): import('./base-adapter').IWorkflowGraph {
    const { SimpleWorkflowGraph } = require('../graph');
    const graph = new SimpleWorkflowGraph();

    graph.addNode('planner', this._plannerNode.bind(this));
    graph.addNode('coder', this._coderNode.bind(this));
    graph.addNode('evaluator', this._evaluatorNode.bind(this));
    graph.addNode('finalize', this._finalizeNode.bind(this));

    graph.setEntryPoint('planner');
    graph.addEdge('planner', 'coder');
    graph.addEdge('coder', 'evaluator');
    graph.addEdge('evaluator', 'finalize');

    return graph;
  }

  // --- Node implementations ---

  private async _plannerNode(state: BaseState): Promise<BaseState> {
    const stateAny = state as Record<string, unknown>;
    const userRequest = (stateAny['user_request'] as string) ?? '';
    const plan = [
      'Analyze requirements and constraints',
      'Implement code changes',
      'Run tests/lint/build/coverage checks',
      'Prepare final output and decision',
    ];
    return { ...state, current_step: 'planner', context: `Code task: ${userRequest}`, implementation_plan: plan } as BaseState;
  }

  private async _coderNode(state: BaseState): Promise<BaseState> {
    const stateAny = state as Record<string, unknown>;
    const metadata = (stateAny['metadata'] as Record<string, unknown>) ?? {};
    let changeSummary = metadata['change_summary'] as string | undefined;
    if (!changeSummary) {
      changeSummary = `Implement request: ${(stateAny['user_request'] as string) ?? ''}`;
    }

    return {
      ...state,
      current_step: 'coder',
      draft_content: changeSummary,
      iteration_count: Number(stateAny['iteration_count'] ?? 0) + 1,
    } as BaseState;
  }

  private async _evaluatorNode(state: BaseState): Promise<BaseState> {
    const evaluation = this.evaluate(state);
    const stateAny = state as Record<string, unknown>;
    const metadata = { ...((stateAny['metadata'] as Record<string, unknown>) ?? {}) };
    metadata['quality_assessment'] = {
      total_score: evaluation.total_score,
      dimension_scores: evaluation.dimension_scores,
    };

    return {
      ...state,
      current_step: 'evaluator',
      decision: evaluation.decision,
      decision_reason: evaluation.decision_reason,
      score: evaluation.total_score,
      feedback_context: evaluation.feedback,
      revision_instructions: evaluation.revision_instructions,
      metadata,
    } as BaseState;
  }

  private async _finalizeNode(state: BaseState): Promise<BaseState> {
    const stateAny = state as Record<string, unknown>;
    const decision = (stateAny['decision'] as string) ?? 'REVISE';
    const approved = decision === 'APPROVED';
    return {
      ...state,
      current_step: 'finalize',
      requires_human_intervention: decision === 'HUMAN_REVIEW',
      final_output: approved
        ? (stateAny['draft_content'] as string) ?? ''
        : (stateAny['final_output'] as string) ?? '',
    } as BaseState;
  }

  // --- Signal resolution helpers ---

  private _resolveBoolSignal(state: BaseState, key: string): boolean | null {
    const stateAny = state as Record<string, unknown>;
    let value: unknown = stateAny[key];

    if (value === undefined) {
      const metadata = (stateAny['metadata'] as Record<string, unknown>) ?? {};
      const qualitySignals = typeof metadata['quality_signals'] === 'object' && metadata['quality_signals'] !== null
        ? (metadata['quality_signals'] as Record<string, unknown>)
        : null;
      if (qualitySignals && key in qualitySignals) {
        value = qualitySignals[key];
      } else if (typeof metadata === 'object' && key in metadata) {
        value = metadata[key];
      }
    }

    return CodeAdapter._toOptionalBool(value);
  }

  private _resolveCoverage(state: BaseState): number | null {
    const stateAny = state as Record<string, unknown>;
    const coverageKeys = ['coverage', 'coverage_pct', 'coverage_percent'];
    const metadata = (stateAny['metadata'] as Record<string, unknown>) ?? {};
    const qualitySignals = typeof metadata['quality_signals'] === 'object' && metadata['quality_signals'] !== null
      ? (metadata['quality_signals'] as Record<string, unknown>)
      : null;

    let value: unknown = undefined;
    for (const key of coverageKeys) {
      if (key in stateAny) { value = stateAny[key]; break; }
      if (qualitySignals && key in qualitySignals) { value = qualitySignals[key]; break; }
      if (typeof metadata === 'object' && key in metadata) { value = metadata[key]; break; }
    }

    if (value === undefined || value === null) return null;
    if (typeof value === 'string') value = value.trim().replace(/%$/, '');
    const coverage = Number(value);
    if (isNaN(coverage)) return null;
    return Math.max(0.0, Math.min(coverage, 100.0));
  }

  private static _toOptionalBool(value: unknown): boolean | null {
    if (value === null || value === undefined) return null;
    if (typeof value === 'boolean') return value;
    if (typeof value === 'number') return Boolean(value);
    if (typeof value === 'string') {
      const n = value.trim().toLowerCase();
      if (['true', 'pass', 'passed', 'yes', 'y', '1'].includes(n)) return true;
      if (['false', 'fail', 'failed', 'no', 'n', '0'].includes(n)) return false;
    }
    return null;
  }

  private static _scoreBinarySignal(signal: boolean | null, passScore: number, unknownScore: number): number {
    if (signal === true) return passScore;
    if (signal === false) return 0.0;
    return unknownScore;
  }

  private static _scoreCoverage(coverage: number | null): number {
    if (coverage === null) return 8.0;
    return Math.round((coverage / 100.0) * 20.0 * 100) / 100;
  }
}

// Auto-register
AdapterRegistry.registerAdapter('code', CodeAdapter, ['strict-governance', 'cli-exposed']);
